go-controller/pkg/ovn/egressgw.go

```diff
@@ -569,41 +569,32 @@ func (oc *Controller) delHybridRoutePolicyForPod(podIP net.IP, node string, forc
 // cleanUpBFDEntry checks if the BFD table entry related to the associated
 // gw router / port / gateway ip is referenced by other routing rules, and if
 // not removes the entry to avoid having dangling BFD entries.
-// This is temporary and can be safely removed when we consume an ovn version
-// that includes http://patchwork.ozlabs.org/project/ovn/patch/3c39dc96a36a3445cfa8485a67de79f9f3d5651b.1614602770.git.lorenzo.bianconi@redhat.com/
-func cleanUpBFDEntry(gatewayIP, gatewayRouter, prefix string) {
+func (oc *Controller) cleanUpBFDEntry(gatewayIP, gatewayRouter, prefix string) {
 	portName := prefix + types.GWRouterToExtSwitchPrefix + gatewayRouter
 
-	output, stderr, err := util.RunOVNNbctl(
-		"--format=csv", "--data=bare", "--no-heading", "--columns=bfd", "find", "Logical_Router_Static_Route", "output_port="+portName, "nexthop=\""+gatewayIP+"\"", "bfd!=[]")
-
-	if err != nil {
-		klog.Errorf("cleanUpBFDEntry: failed to list routes for %s, stderr: %q, (%v)", portName, gatewayIP, err, stderr)
-		return
-	}
-	// the bfd entry is still referenced, meaning there's another route on the router
-	// referencing it.
-	if strings.TrimSpace(output) != "" {
-		return
-	}
-	uuids, stderr, err := util.RunOVNNbctl(
-		"--format=csv", "--data=bare", "--no-heading", "--columns=_uuid", "find", "BFD", "logical_port="+portName, "dst_ip=\""+gatewayIP+"\"")
+	logicalRouterStaticRouteRes := []nbdb.LogicalRouterStaticRoute{}
+	err := oc.nbClient.WhereCache(func(lrsr *nbdb.LogicalRouterStaticRoute) bool {
+		return lrsr.OutputPort != nil && *lrsr.OutputPort == portName && lrsr.Nexthop == gatewayIP && lrsr.BFD != nil && *lrsr.BFD != ""
+	}).List(&logicalRouterStaticRouteRes)
 	if err != nil {
-		klog.Errorf("Failed to list routes for %s, stderr: %q, (%v)", gatewayRouter, err, stderr)
+		klog.Errorf("cleanUpBFDEntry: failed to list routes for %s, err: %v", portName, err)
 		return
 	}
 
-	if strings.TrimSpace(uuids) == "" {
-		klog.Infof("Did not find bfd entry for %s %s", portName, gatewayIP)
+	if len(logicalRouterStaticRouteRes) > 0 {
 		return
 	}
 
-	for _, uuid := range strings.Split(uuids, "\n") {
-		_, stderr, err = util.RunOVNNbctl("--if-exists", "destroy", "BFD", uuid)
-		if err != nil {
-			klog.Errorf("Failed to destroy BFD %s, stderr: %q, (%v)",
-				uuid, stderr, err)
-		}
+	opModels := []libovsdbops.OperationModel{
+		{
+			ModelPredicate: func(bfd *nbdb.BFD) bool {
+				return bfd.LogicalPort == portName && bfd.DstIP == gatewayIP
+			},
+			ExistingResult: &[]nbdb.BFD{},
+		},
+	}
+	if err := oc.modelClient.Delete(opModels...); err != nil {
```

Comment:
@alexanderConstantinescu : so here, do we query the BFD libovsdb cache to check if the bfd exists with that match first? Or are we triggering a direct delete transaction on the db irrespective of whether an entry exists or not? Seems like we do the latter here... wondering which would be more efficient....

Reply:
> Seems like we do the latter here 

No, we are doing the former. `oc.modelClient.Delete` is the equivalent of `ovn-nbctl --if-exists remove/delete`, in this case we retrieve all items found by the predicate and store them in `ExistingResult`, we then proceed to deleting those existing items.

Reply:
ok so the `ModelPredicate` is really just querying the bfd libovsdb cache right? `ovn-nbctl --if-exists remove/delete` is still a direct call to the db. As long as we don't directly go to db and are checking the cache first which is what you are saying, we are good.